src/compiler/checker.ts

```diff
@@ -15568,6 +15574,32 @@ namespace ts {
             return getReturnTypeOfSignature(signature);
         }
 
+        function checkImportCallExpression(node: ImportCall): Type {
+            // Check grammar of dynamic import
+            if (checkGrammarArguments(node, node.arguments) || checkGrammarImportCallExpression(node)) {
+                return createPromiseReturnType(node, anyType);
+            }
+
+            if (modulekind === ModuleKind.ES2015) {
+                grammarErrorOnNode(node, Diagnostics.Dynamic_import_cannot_be_used_when_targeting_ECMAScript_2015_modules);
+            }
+
+            const specifier = node.arguments[0];
+            const specifierType = checkExpression(specifier);
+            if (!isTypeAssignableTo(specifierType, stringType)) {
+                error(specifier, Diagnostics.Dynamic_import_s_specifier_must_be_of_type_string_but_here_has_type_0, typeToString(specifierType));
+            }
+            // resolveExternalModuleName will return undefined if the moduleReferenceExpression is not a string literal
+            const moduleSymbol = resolveExternalModuleName(node, specifier);

```

Comment:
Nop. WE only do when it is strictly literal type. with union of literal types we will just treat it as any. Though now that I consider it, we should do union literal type and just return union of resolve module